src/overlay/FlowControlCapacity.cpp

```diff
@@ -0,0 +1,187 @@
+// Copyright 2022 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "overlay/FlowControlCapacity.h"
+#include "main/Application.h"
+#include "overlay/OverlayManager.h"
+#include "util/Logging.h"
+#include <Tracy.hpp>
+
+namespace stellar
+{
+
+FlowControlCapacity::FlowControlCapacity(Application& app, NodeID nodeID)
+    : mApp(app), mNodeID(nodeID)
+{
+}
+
+void
+FlowControlCapacity::checkCapacityInvariants() const
+{
+    ZoneScoped;
+    auto limits = getCapacityLimits();
+    releaseAssert(getCapacityLimits().mFloodCapacity >=
+                  mCapacity.mFloodCapacity);
+    if (limits.mTotalCapacity)
+    {
+        releaseAssert(mCapacity.mTotalCapacity);
+        releaseAssert(*limits.mTotalCapacity >= *mCapacity.mTotalCapacity);
+    }
+    else
+    {
+        releaseAssert(!mCapacity.mTotalCapacity);
+    }
+}
+
+void
+FlowControlCapacity::lockOutboundCapacity(StellarMessage const& msg)
+{
+    ZoneScoped;
+    if (mApp.getOverlayManager().isFloodMessage(msg))
+    {
+        releaseAssert(hasOutboundCapacity(msg));
+        mOutboundCapacity -= getMsgResourceCount(msg);
+    }
+}
+
+bool
+FlowControlCapacity::lockLocalCapacity(StellarMessage const& msg)
+{
+    ZoneScoped;
+    checkCapacityInvariants();
+    auto msgResources = getMsgResourceCount(msg);
+    if (mCapacity.mTotalCapacity)
+    {
+        releaseAssert(mCapacity.mTotalCapacity >= msgResources);
+        *mCapacity.mTotalCapacity -= msgResources;
+    }
+
+    if (mApp.getOverlayManager().isFloodMessage(msg))
+    {
+        // No capacity to process flood message
+        if (mCapacity.mFloodCapacity < msgResources)
+        {
+            return false;
+        }
+
+        mCapacity.mFloodCapacity -= msgResources;
+        if (mCapacity.mFloodCapacity == 0)
+        {
+            CLOG_DEBUG(Overlay, "No flood capacity for peer {}",
+                       mApp.getConfig().toShortString(mNodeID));
+        }
+    }
+
+    return true;
+}
+
+uint64_t
+FlowControlCapacity::releaseLocalCapacity(StellarMessage const& msg)
+{
+    ZoneScoped;
+    uint64_t releasedFloodCapacity = 0;
+    size_t resourcesFreed = getMsgResourceCount(msg);
+    if (mCapacity.mTotalCapacity)
+    {
+        *mCapacity.mTotalCapacity += resourcesFreed;
+    }
+
+    if (mApp.getOverlayManager().isFloodMessage(msg))
+    {
+        if (mCapacity.mFloodCapacity == 0)
+        {
+            CLOG_DEBUG(Overlay, "Got flood capacity for peer {} ({})",
```

Comment:
+1 adding to both total capacity and flood capacity, and debug log when will make flood capacity >=0